File: leetcode/python/1457_pseudo-palindromic_paths_in_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pseudoPalindromicPaths (self, root: TreeNode | None) -> int:
        if not root:
            return 0
        bucket = [0] * 10
        res: int = 0
        def dfs(root: TreeNode) -> None:
            bucket[root.val] += 1
            nonlocal res
            if not root.left and not root.right:
                res += self.isPseudoPalindrome(bucket)
            if root.left:
                dfs(root.left)
            if root.right:
                dfs(root.right)
            bucket[root.val] -= 1
        dfs(root)
        return res

    def isPseudoPalindrome(self, bucket: list[int]) -> bool:
        odds: int = 0
        for count in bucket:
            odds += count & 1
            if odds == 2:
                return False
        return True
        

File: leetcode/rust/src/1457_pseudo-palindromic_paths_in_binary_tree.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn pseudo_palindromic_paths (root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if root.is_none() {
            return 0;
        }

        let mut bucket = [0; 10];
        Self::dfs(&root, &mut bucket)
    }

    fn dfs(root: &Option<Rc<RefCell<TreeNode>>>, bucket: &mut [i32; 10]) -> i32 {
        if let Some(node) = root {
            let node = node.borrow();
            bucket[node.val as usize] += 1;
            let mut count = 0;
            if node.left.is_none() && node.right.is_none() {
                if Self::is_pseudo_palindrome(bucket) {
                    count += 1;
                }
            } else {
                count += Self::dfs(&node.left, bucket);
                count += Self::dfs(&node.right, bucket);
            }
            bucket[node.val as usize] -= 1;
            return count;
        }
        0
    }

    fn is_pseudo_palindrome(bucket: &[i32; 10]) -> bool {
        let mut odd_count = 0;
        for i in 0..10 {
            if bucket[i] % 2 == 1 {
                odd_count += 1;
            }
        }
        odd_count <= 1
    }
}